enhancements/cluster-validation-tool.md

```diff
@@ -0,0 +1,521 @@
+    enabled: true                                             |      enabled: false
+
+---
+<next CR>
+...
+
+Summary
+  Missing required CRs
+    someCR
+    anotherCR
+  Unmatched CRs
+    unknownCR
+    anotherUnknownCR
+  Total CRs with drift: 2
+```
+
+#### Tool delivery
+
+The tool will be a standalone binary executable suitable for direct use by customers, developers,
+support engineering, etc. The tool may be installed as a kubectl/oc plugin.
+
+### Reference Configuration Specification
+
+The reference configuration consists of a set of CRs (yaml files with one CR per file) with
+annotations/metadata to describe expected variability. The reference configuration CRs conform to
+these criteria:
+1. Each CR may be annotated as a required (ie a compliant cluster must contain at least one of that
+   CR) or optional CR.
+1. Each CR may contain annotations to indicate fields which have expected user variation or which
+   are optional.
+1. Any field of a reference configuration CR which is not otherwise annotated is required and the
+   value must be as specified in order to be compliant.
+
+#### Reference CR Annotations
+
+Annotations attached to the CRs are used for providing additional data used during validation. In
+the initial version of this tool the annotation is used in making two determinations:
+1. The `component-name` is used to group together related CRs in order to capture optional groups of
+   CRs (eg combination of optional Operator subscription and configuration CRs). If any of the
+   required CRs (see `required` next) are included in the input configuration then all required CRs
+   in the group are expected to be included and any which are missing will be reported. If none of
+   the required CRs in the group are included then no report of "missing content" for the group will
+   be generated. Optionally the tool may allow for explicit include/exclude of a component-name by
+   command line argument.
+1. The `required` field indicates if the CR must be present in the input configuration. Any required
+   CR which is not included is reported in the summary as "missing content". The required field has
+   a scope of "within the component-name". If a specfici component-name is not included in the
+   analysis then any required CRs within that component are not flagged as missing content.
+
+| key | value | default |
+| --- | ---- | --- |
+| part-of | the highest level that the CR is part-of e.g DU. | "" |
+| required | must be available. | true |
+| component-name | bundle up sibling CRs. E.g ref PerformaceProfile and TuneD. | "" |
+
+```yaml
+metadata:
+  annotations:
+    reference-config.openshift.io/attributes: |
+      {"part-of": "sample.reference.config",
+       "required": true,
+       "component-name": "someSubComponent"
+       }
+```
+
+#### Example Reference Configuration CR
+User variable content is handled by golang formatted templating within the reference configuration
```

Comment:
Are you reinventing kustomize here? Is it a good idea to integrate templating vs. use external templating and let the diff tool do just the diffing?

Reply:
The use of templating here is intended to give enough structure and functionality to capture a range of complexity in reference configurations. The templating allows the reference to communicate expected user variation, rules around acceptable/verified values, and optional content (which if included must fall within acceptable bounds as well). Templating languages have a couple of advantages:
* There is already common use of them within the domain of k8s/ocp clusters
* The languages are well understood, documented, and supported by various libraries. 
* They are feature rich and support extension when needed

We definitely aren't intending to reinvent kustomize, helm or any other deployment oriented tool flow. In this case the templating would be used within the validation tool to pull select expected user variable content from the user's input CRs directly into a "rendered" version of the reference configuration. This results in a CR that can be diff'd against the original input CR using standard diff tools/techniques. This happens within the context of the validation tool which is able to analyze both the input CRs and the reference configuration in order to do the correlation (which files get compared against one another). Once correlated the rendering is done (handling all user variable content) and then diff'd. The actual step of template processing is fairly trivial due to the library support available.

I agree that the reference configuration should be as human readable as possible, but that it should also allow:
* enough structure to capture variations, validations, groups of optional content (with required values within)
* flexibility to address future reference configurations without updating/modifying the validation tool
* other tools/toolchains to be built around the reference (by using a standardized format for the notations)